File: 卡码/代码随想录/1/09-动态规划/Go/04-股票问题/LeetCode_0122.go
// 122. 买卖股票的最佳时机 II
package main

import (
	"math"
)

func maxProfit(prices []int) int {
	dp := []int{-prices[0], 0}
	for i := 1; i < len(prices); i++ {
		dp[0] = int(math.Max(float64(dp[0]), float64(dp[1]-prices[i])))
		dp[1] = int(math.Max(float64(dp[1]), float64(dp[0]+prices[i])))
	}
	return dp[1]
}

File: 卡码/代码随想录/Java/codeThoughts/src/main/java/com/kurumi/leetCode/codeThoughts/动态规划/_04_股票问题/LeetCode_0122.java
package com.kurumi.leetCode.codeThoughts.动态规划._04_股票问题;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 122. 买卖股票的最佳时机 II
 */
public class LeetCode_0122 {
    static class Solution {
        public int maxProfit(int[] prices) {
            int[] dp = new int[] { -prices[0], 0 };
            for (int i = 1; i < prices.length; i++) {
                dp[0] = Math.max(dp[0], dp[1] - prices[i]);
                dp[1] = Math.max(dp[1], dp[0] + prices[i]);
            }
            return dp[1];
        }
    }
}
